File: 2024/day_06/part_1.lua
-- https://adventofcode.com/2024/day/6

local function read_input_file(filename)
  local dir = debug.getinfo(1, 'S').source:sub(2):match('^(.*)/')
  local file, err
  if dir == nil then
    file, err = io.open(filename)
  else
    file, err = io.open(dir .. '/' .. filename)
  end
  if not file then
    print('input opening failed ' .. err)
    os.exit()
  end
  return file
end

local function find_in_table(tab, target)
  if tab == nil then
    return 0
  end
  for i = 1, #tab do
    if tab[i] == target then
      return i
    end
  end
  return 0
end

local function is_edge(map, row, col)
  return row == 1 or row == #map or col == 1 or col == string.len(map[1])
end

local function main()
  local result = 0
  local map = {}

  local file = read_input_file('input.txt')
  local row, col = nil, nil
  local line_num = 1
  for line in file:lines() do
    table.insert(map, line)
    if row == nil then
      local spos, _ = string.find(line, '^', 1, true)
      if spos ~= nil then
        row = line_num
        col = spos
      end
    end
    line_num = line_num + 1
  end
  file:close()

  local dr, dc = -1, 0
  local visited = {}
  while not is_edge(map, row, col) do
    if find_in_table(visited[row], col) == 0 then
      result = result + 1
      if visited[row] == nil then
        visited[row] = { col }
      else
        table.insert(visited[row], col)
      end
    end
    if map[row + dr]:sub(col + dc, col + dc) == '#' then
      dr, dc = dc, -dr
    end
    row = row + dr
    col = col + dc
  end

  print(result + 1)
end

main()

File: 2024/day_10/part_1.py
""" https://adventofcode.com/2024/day/10 """

import os


def get_file_path(filename: str) -> str:
    dir_path = os.path.split(os.path.abspath(__file__))[0]
    return os.path.join(dir_path, filename)


def search(
    row: int, col: int, grid: list[list[int]], visited: set[tuple[int, int]]
) -> int:
    if (row, col) in visited:
        return 0
    visited.add((row, col))
    if grid[row][col] == 9:
        return 1
    score = 0
    for dr, dc in [(-1, 0), (1, 0), (0, 1), (0, -1)]:
        nr, nc = row + dr, col + dc
        if nr < 0 or nc < 0 or nr >= len(grid) or nc >= len(grid[nr]):
            continue
        if grid[nr][nc] - grid[row][col] == 1:
            score += search(nr, nc, grid, visited)
    return score


def main():
    grid: list[list[int]]

    with open(get_file_path("input.txt")) as fd:
        grid = [list(map(int, list(line.rstrip()))) for line in fd.readlines()]

    result = 0
    visited: set[tuple[int, int]] = set()
    for i in range(len(grid)):
        for j in range(len(grid[i])):
            if grid[i][j] == 0:
                visited.clear()
                result += search(i, j, grid, visited)

    print(result)


main()
